java/statestore/src/test/java/sleeper/statestore/StateStoreProviderWithSizeTest.java

```diff
@@ -0,0 +1,115 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.statestore;
+
+import org.junit.jupiter.api.Test;
+
+import sleeper.configuration.properties.instance.InstanceProperties;
+import sleeper.configuration.properties.table.TableProperties;
+import sleeper.core.schema.Schema;
+import sleeper.core.statestore.StateStore;
+import sleeper.core.statestore.inmemory.StateStoreTestHelper;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static sleeper.configuration.properties.InstancePropertiesTestHelper.createTestInstanceProperties;
+import static sleeper.configuration.properties.instance.CommonProperty.STATESTORE_PROVIDER_CACHE_SIZE;
+import static sleeper.configuration.properties.table.TablePropertiesTestHelper.createTestTableProperties;
+import static sleeper.configuration.properties.table.TableProperty.TABLE_ID;
+import static sleeper.configuration.properties.table.TableProperty.TABLE_NAME;
+import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;
+
+public class StateStoreProviderWithSizeTest {
+    private final InstanceProperties instanceProperties = createTestInstanceProperties();
+    private final Schema schema = schemaWithKey("key");
+    private final Map<String, StateStore> tableIdToStateStore = new HashMap<>();
+    private final List<String> tablesLoaded = new ArrayList<>();
+
+    @Test
+    void shouldCacheStateStore() {
+        // Given
+        TableProperties table = createStateStore("test-table-id", "test-table");
+
+        // When
+        Stream<StateStore> stateStores = providerGettingStores(table, table, table);
+
+        // Then
+        assertThat(stateStores).containsExactly(
+                stateStore(table),
+                stateStore(table),
+                stateStore(table));
+        assertThat(tablesLoaded).containsExactly("test-table-id");
+    }
+
+    @Test
+    void shouldRemoveOldestCacheEntryWhenLimitHasBeenReached() {
+        // Given
+        instanceProperties.setNumber(STATESTORE_PROVIDER_CACHE_SIZE, 2);
+        TableProperties table1 = createStateStore("table-id-1", "test-table-1");
+        TableProperties table2 = createStateStore("table-id-2", "test-table-2");
+        TableProperties table3 = createStateStore("table-id-3", "test-table-3");
+
+        // When
+        Stream<StateStore> stateStores = providerGettingStores(
+                table1,
+                table2,
+                // Cache size has been reached, oldest state store will be removed from cache
+                table3,
+                table1);
```

Comment:
I'm finding it hard to see what this is doing now. Can we remove the stream and put it back to call the method directly 4 times?